runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/ReferenceCountingFlinkExecutableStageContextFactory.java

```diff
@@ -104,9 +108,21 @@ private void scheduleRelease(JobInfo jobInfo) {
     WrappedContext wrapper = getCache().get(jobInfo.jobId());
     Preconditions.checkState(
         wrapper != null, "Releasing context for unknown job: " + jobInfo.jobId());
-    // Do not release this asynchronously, as the releasing could fail due to the classloader not being
-    // available anymore after the tasks have been removed from the execution engine.
-    release(wrapper);
+
+    PipelineOptions pipelineOptions =
+        PipelineOptionsTranslation.fromProto(jobInfo.pipelineOptions());
+    int environmentCacheTTLMillis =
+        pipelineOptions.as(PortablePipelineOptions.class).getEnvironmentCacheMillis();
+    if (environmentCacheTTLMillis > 0) {
+      // Schedule task to clean the container later.
+      // Ensure that this class is loaded in the parent Flink classloader.
+      getExecutor()
+          .schedule(() -> release(wrapper), environmentCacheTTLMillis, TimeUnit.MILLISECONDS);
```

Comment:
These changes are already contained in #6638. You could base this PR on #6638.

Reply:
Yes, that's one of those situations where 2 separate commits in the same PR make sense. But since you have the other PR under review already, why not wrap that up first.

Reply:
> There is no way to make PR depend upon other PR in git so had to resort to adding 2 separate commits.

Not true, you can just specify the branch of the other PR as base branch.

Reply:
Maybe I am missing some thing.try t
For this PR, I based my branch java_pvr_jenkins on  java_pvr_cache_environments and then created the PR (merge 1 commit into apache:master from angoenka:java_pvr_jenkins).
This picks the commits from both branches java_pvr_jenkins and java_pvr_cache_environments.
Is there anyway to avoid picking commits from the base PR?

Reply:
You can base a PR on another PR's branch. Not sure how much sense that makes because then you would merge into your fork instead of the upstream repo. So I take that back :)